server/src/main/java/org/opensearch/gateway/GatewayAllocator.java

```diff
@@ -303,6 +313,59 @@ protected AsyncShardFetch.FetchResult<TransportNodesListGatewayStartedShards.Nod
         }
     }
 
+
+    class InternalPrimaryBatchShardAllocator extends PrimaryShardBatchAllocator {
+
+        @Override
+        @SuppressWarnings("unchecked")
+        protected AsyncShardFetch.FetchResult<TransportNodesListGatewayStartedShardsBatch.NodeGatewayStartedShardsBatch> fetchData(Set<ShardRouting> shardsEligibleForFetch,
+                                                                                                                                   Set<ShardRouting> inEligibleShards,
+                                                                                                                                   RoutingAllocation allocation) {
+            ShardRouting shardRouting = shardsEligibleForFetch.iterator().hasNext() ? shardsEligibleForFetch.iterator().next() : null;
+            shardRouting = shardRouting == null && inEligibleShards.iterator().hasNext() ? inEligibleShards.iterator().next() : shardRouting;
+            if (shardRouting == null) {
+                return new AsyncShardFetch.FetchResult<>(null, Collections.emptyMap());
```

Comment:
Let's use `DiscoveryNodes.EMPTY_NODES` instead  of passing `null` for `DiscoveryNodes`. This will ensure we dont see NPE at https://github.com/opensearch-project/OpenSearch/pull/8742/files#diff-ed0f7edad0065bfd401b62572d9e6d20bfd2d0476d5e5f75918b684cb2ca59e3R176

Additionally, please make sure this case is covered with a UT.

Reply:
No we cant do that. FetchResult first parameter is not of type DiscoveryNodes but rather a map that  is the response that transport layer sends and that transport is decided by PSBA/RSBA. They are consumers of this response. They have the checks in place for null [here](https://github.com/opensearch-project/OpenSearch/pull/8916/files#diff-22046490171b16573182a4088e32b54d06329c70fbf48af6d19e3bb4ac8a6914R116)

Also we already have this [documentation](https://github.com/Gaurav614/OpenSearch/blob/7dc30a4fc8744146fb387971bbf6af959956f3b5/server/src/main/java/org/opensearch/gateway/AsyncShardFetch.java#L419) in place to make hasData() call before making any get data call. Consumers have those check in place

Reply:
I am looking at https://github.com/opensearch-project/OpenSearch/pull/8742/files#diff-ed0f7edad0065bfd401b62572d9e6d20bfd2d0476d5e5f75918b684cb2ca59e3R164, Are we changing the method signature for `FetchResult` and getting rid of the first param `DiscoveryNodes nodes` ?  Can you link the PR here. 

```
public synchronized FetchResult<T> fetchData(DiscoveryNodes nodes, 
    Map<ShardId, Set<String>> ignoreNodes) {
```

Reply:
I am looking at https://github.com/opensearch-project/OpenSearch/pull/8742/files#diff-ed0f7edad0065bfd401b62572d9e6d20bfd2d0476d5e5f75918b684cb2ca59e3R164, Are we changing the method signature for `FetchResult` and getting rid of the first param `DiscoveryNodes nodes` ?  Can you link the PR here. 

```
public synchronized FetchResult<T> fetchData(DiscoveryNodes nodes, 
    Map<ShardId, Set<String>> ignoreNodes) {
```